### PR TITLE
Add long error diagnostics for E0401

### PR DESCRIPTION
**Pull request by Manishearth**

### CODE REVIEW

File: src/librustc_resolve/diagnostics.rs [hunk 1]
```diff
@@ -273,6 +273,77 @@ on this topic:
...
+    fn bar<T: Copy>(y: T) {
+        // ..
+    }
+    bar(x);
+}
+```
+
+This may require additional type hints in the function body.
+
+In case the function is in an `impl`, defining a private helper function might
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: impl block maybe ?

Comment (Manishearth) [inline src/librustc_resolve/diagnostics.rs]: not really a block

Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: How who you call it ?

Comment (Manishearth) [inline src/librustc_resolve/diagnostics.rs]: I just call it an impl

Comment (steveklabnik) [inline src/librustc_resolve/diagnostics.rs]: I hear people say both.

**Pull request merged**